File: h2o-docs/src/booklets/v2_2015/source/deeplearning/deeplearning_anomaly.R
# TO DO: Check this code

# Upload ECG train and test data
train_ecg <- h2o.importFile(path = "ecg_train.csv", 
                            header = FALSE, 
                            sep = ",", 
                            key = "train_ecg") 
test_ecg <- h2o.importFile(path = "ecg_test.csv", 
                           header = FALSE, 
                           sep = ",", 
                           key = "test_ecg") 

# Train deep autoencoder learning model on "normal" 
# training data, y ignored 
anomaly_model = h2o.deeplearning(x = 1:210, 
                                 y = 1, 
                                 valdiation_frame = train_ecg, 
                                 activation = "Tanh", 
                                 autoencoder = T, 
                                 hidden = c(50,20,50), 
                                 l1 = 1E-4, 
                                 epochs = 100)                 

# Compute reconstruction error with the Anomaly 
# detection app (MSE between output layer and input layer)
recon_error <- h2o.anomaly(test_ecg, anomaly_model)

# Pull reconstruction error data into R and 
# plot to find outliers (last 3 heartbeats)
recon_error <- as.data.frame(recon_error)
recon_error
plot.ts(recon_error)

# Note: Testing = Reconstructing the test dataset
test_recon <- h2o.predict(anomaly_model, test_ecg) 
head(test_recon)

File: h2o-docs/src/booklets/v2_2015/source/deeplearning/deeplearning_anomaly.py
# Download and import ECG train and test data into the H2O cluster
train_ecg = h2o.import_file("http://h2o-public-test-data.s3.amazonaws.com/smalldata/anomaly/ecg_discord_train.csv")
test_ecg = h2o.import_file("http://h2o-public-test-data.s3.amazonaws.com/smalldata/anomaly/ecg_discord_test.csv")


# Train deep autoencoder learning model on "normal" 
# training data, y ignored
anomaly_model = h2o.deeplearning(
        x=train_ecg.names,  
        training_frame=train_ecg, 
        activation="Tanh", 
        autoencoder=True,
        hidden=[50,50,50], 
        l1=1e-4, 
        epochs=100)                

# Compute reconstruction error with the Anomaly 
# detection app (MSE between output layer and input layer)
recon_error = anomaly_model.anomaly(test_ecg)


# Note: Testing = Reconstructing the test dataset
test_recon = anomaly_model.predict(test_ecg) 
test_recon
